Hbase uppercase table

aimethed opened the pull request:
*Description of changes:*
Previously, hbase would fail when trying to access tables with uppercase characters because athena does not support uppercase tables. This change allows querying of hbase uppercase tables by handling uppercase names the same way the dynamodb connector handles this: by getting all tables in the hbase namespace and mapping the normal case name to its lowercase name. This ensures that if multiple hbase tables would have the same lowercase name (ex: Table and tAble would both be table), an error is thrown.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

chngpe commented on athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseTableNameUtils.java:
```diff
@@ -0,0 +1,152 @@
+    public static org.apache.hadoop.hbase.TableName getHbaseTableName(HBaseConnection conn, TableName athTableName)
+            throws IOException
+    {
+        if (conn.tableExists(getQualifiedTable(athTableName))) {
```
Double check, we lower case the athTableName here. If Hbase has Uppercase, conn.getTableExists will able to find that table?

aimethed replied:
This is a check to see if the table we are trying to get is already lowercase. If it is (tableExists), then we can skip the rest of the logic. Otherwise, we run the caseInsensitiveSearch.

chngpe commented on athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseMetadataHandler.java:
```diff
@@ -197,7 +198,7 @@ public ListTablesResponse doListTables(BlockAllocator blockAllocator, ListTables
             TableName tableName = tables[i];
             tableNames.add(new com.amazonaws.athena.connector.lambda.domain.TableName(request.getSchemaName(),
-                    tableName.getNameAsString().replace(request.getSchemaName() + ":", "")));
+                    tableName.getNameAsString().toLowerCase(Locale.ENGLISH).replace(request.getSchemaName() + ":", "")));
```
can we not do this? so metadata can reflect correctly

aimethed replied:
If the tableName has uppercase characters, not lower casing here would cause an error as Athena TableNames don't support upper case characters.

aimethed replied:
Just called, trino will lowercase it, so it is not needed here. Changing

AbdulR3hman commented on athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseTableNameUtils.java:
```diff
@@ -0,0 +1,152 @@
+            return caseInsensitiveMatch.get();
+        }
+        else {
+            return null;
```
I think its a better practice to change this to Optional.Empty();

aimethed replied:
Just updated the code so no longer has a null return. Now throws an error for 0 matches to mimic docdb behavior

AbdulR3hman commented on athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseTableNameUtils.java:
```diff
@@ -0,0 +1,152 @@
+     * @throws IOException 
+     */
+    @VisibleForTesting
+    protected static Optional<org.apache.hadoop.hbase.TableName> tryCaseInsensitiveSearch(HBaseConnection conn, TableName tableName)
```
I would add for the reader of this function is that HBase actually supports case sensitive tables; meaning two tables with the same name can co-exist. And we are attempting to resolve case insensitive `only and only if` there is one table per that string.

aimethed replied:
I think that is covered with "To prevent ambiguity, an IllegalStateException is thrown if multiple tables map to the given tableName."

AbdulR3hman commented on athena-hbase/src/main/java/com/amazonaws/athena/connectors/hbase/HbaseTableNameUtils.java:
```diff
@@ -0,0 +1,152 @@
+            lowerCaseNameMapping.put(nextTableName.getQualifierAsString().toLowerCase(Locale.ENGLISH), nextTableName.getNameAsString());
+        }
+        Collection<String> mappedNames = lowerCaseNameMapping.get(tableName.getTableName());
+        if (mappedNames.size() > 1) {
```
this needs to be tested; I am worried that an exception will bubble up to the console when loading the schema for the database which contain two tables with the same string (different capitalization)

aimethed replied:
This exception does cause the query to fail (it was tested). This is the intended behavior.

AbdulR3hman commented:
I believe we should preserve the original behavior; and put this change behind a feature flag as it will change what existing customers are expecting the default connector behavior.

aimethed commented:
Summary of changes:

- Now requires the following environment variable to be added to enable case insensitive search: `enable_case_insensitive_match` = `true`
- Updated doGetTable to return the name from the case insensitive search instead of the one from the request
  - If it's disabled, just returns the name
- Removed the check for if the table exists. Since this feature is now opt-in, the check causes unwanted behavior.
  - If there are two tables, `table` and `Table`, querying for `Table` before would succeed but match with `table`. Now errors due to multiple possible matches.
- Added a check for if the given name isn't lowercase. Addresses an error where doListTables was called with correctly-cased table name.
